plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -158,28 +178,36 @@ floatTerm ver t0 = do
     floatTermInner usgs = go
       where
         -- Float bindings in the given `Term` inwards, and annotate each term with the set of
-        -- used term `Unique`s in the `Term`.
+        -- `Unique`s of used variables in the `Term`.
         go ::
             Term tyname name uni fun a ->
             Term tyname name uni fun (a, Uniques)
         go t = case t of
             Apply a fun0 arg0 ->
                 let fun = go fun0
                     arg = go arg0
-                    us = termUniqs fun `Set.union` termUniqs arg
+                    us = termUniqs fun <> termUniqs arg
                  in Apply (a, us) fun arg
```

Comment:
This does all feel like a lot of boilerplate to me and I do wonder if there's a nice way to do a bunch of this generically with some traversals or something :thinking:

Reply:
Well, or we have lots of `Traversals` which are IMO better https://www.michaelpj.com/blog/2020/08/02/lenses-for-tree-traversals.html